### PR TITLE
Introduce kj::List.

### PR DESCRIPTION
**Pull request by kentonv**
This is a pattern I've repeated a million times in KJ and beyond, it's time to formalize it.

### CODE REVIEW

File: c++/src/kj/list.h [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+  // if the List is guarded by a mutex, and that mutex is not currenty locked. Normally, you should
+  // have T's destructor remove it from any lists. You can use `link.isLinked()` to check if the
+  // item is currently in a list.
+  //
+  // kj::List is a doubly-linked list in order to allow O(1) removal of any element given only a
+  // reference to the element. However, it only supports forward iteration.
+  //
+  // When iterating over a kj::List, you can safely remove current element which the iterator
+  // points to without breaking the iteration. However, removing any *other* element could
+  // invalidate the iterator.
```
Comment (harrishancock) [inline c++/src/kj/list.h:83]: This behavior would be good to test.

**Review by harrishancock — APPROVED** [review]
I love it! :D

**Comment by xortive** [comment]
nice!

File: c++/src/kj/list.h (hunk 1 above)
Comment (kentonv) [reply c++/src/kj/list.h:83]: done

**Pull request merged**